File: src/main/java/br/com/rfasioli/micrometermetrics/messaging/stream/PersonProducer.java
package br.com.rfasioli.micrometermetrics.messaging.stream;

import br.com.rfasioli.micrometermetrics.resource.dto.PersonDTO;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Sinks;

import java.util.function.Supplier;

@Component
@RequiredArgsConstructor
public class PersonProducer implements Supplier<Flux<PersonDTO>> {
  private final Sinks.Many<PersonDTO> sink = Sinks.many().unicast().onBackpressureBuffer();

  @Override
  public Flux<PersonDTO> get() {
    return sink.asFlux();
  }

  public void produce(final PersonDTO personDTO) {
    sink.emitNext(personDTO, Sinks.EmitFailureHandler.FAIL_FAST);
  }
}

File: src/main/kotlin/br/com/rfasioli/micrometermetrics/messaging/stream/PersonProducer.kt
package br.com.rfasioli.micrometermetrics.messaging.stream

import br.com.rfasioli.micrometermetrics.resource.dto.PersonDTO
import org.springframework.stereotype.Component
import reactor.core.publisher.Flux
import reactor.core.publisher.Sinks
import java.util.function.Supplier

@Component
class PersonProducer : Supplier<Flux<PersonDTO>> {
    private val sink = Sinks.many().unicast().onBackpressureBuffer<PersonDTO>()

    override fun get() = sink.asFlux()

    fun produce(personDTO: PersonDTO) {
        sink.tryEmitNext(personDTO).orThrow()
    }
}
